Build system enhancements

shindere opened the pull request:
In preparation for both makefiles refactoring (to get more parallelism
during build) and autoconf migration.

Cc @xavierleroy @alainfrisch @dra27 @adrien-n

shindere committed:
Fix NATIVECC's value for the MSVC64 port

The -D_CRT_SECURE_NO_DEPRECATE argument was not given to NATIVECC,
this seems to be an omission.

shindere committed:
configure system fixup

For the verbose variable, use boolean values true and false rather
than yes and no.

Also fix indentation in scripts under config/auto-aux where appropriate.

shindere committed:
A bit of code refactoring in the scripts under config/auto-aux

shindere committed:
Get rid of the NATIVECCLINKOPTS build variable.

This variable is empty and actually never used in the build system.

shindere committed:
Rename the BYTECCLINKOPTS build variable to LDFLAGS

shindere committed:
Configure helper scripts enhancement

When calling the C compiler, the scripts under config/auto-aux now
take into account the value of the cflags and ldflags environment variables.

This will make it possible to invoke these scripts from configure in a more
straightforward way.

shindere committed:
configure: use cflags and ldflags when calling helper scripts

shindere committed:
Use only one variable for rpath

Before this commit there was BYTECCRPATH and NATIVECCRPATH, but they
were actually identical to each other.

This commit gets rid of them and uses the RPATH variable.

shindere committed:
Get rid of BYTECC and NATIVECC

Use the same C compiler, CC, in all the build system.

shindere committed:
configure script enhancement

Let the with_sharedlibs variable take boolean values rather than yes
and no. This simplifies tests a bit.

shindere committed:
Define new build variables for C compiler and preprocessor flags

shindere committed:
Introduce and use the OUTPUTEXE and OUTPUTOBJ build variables

adrien-n commented on configure:
```diff
@@ -323,6 +323,10 @@ fi
 
+# Configure compiler to use in further tests.
+
+export cc="$cc"
```
You can/should write only

    export cc

More portable (even though we don't care about systems with shells that don't handle this) and consistent.

adrien-n commented on configure:
```diff
@@ -951,17 +944,11 @@ case "$native_compiler" in
       arch=none; model=default; system=unknown; natdynlink=false;;
 esac
 
-if test -z "$ccoption"; then
```
To be sure I got it right and for other reviewers, this block is removed because it duplicates functionality which is around line 310 and already sets $cc according to $ccoption. Is that right?

adrien-n commented on testsuite/tests/lib-dynlink-native/Makefile:
```diff
@@ -108,7 +108,7 @@ mylib.cmxa: plugin.cmx plugin2.cmx
 
 factorial.$(O): factorial.c
-	@$(OCAMLOPT) -c -ccopt "$(SHAREDCCCOMPOPTS)" -ccopt -I \
+	$(OCAMLOPT) -verbose -c -ccopt "$(SHAREDCCCOMPOPTS)" -ccopt -I \
```
You've lost the '@' at the beginning of the command-line.

adrien-n commented on config/Makefile.msvc:
```diff
@@ -97,12 +97,11 @@ AFL_INSTRUMENT=false
+CFLAGS = -nologo -O2 -Gy- -MD
+CPPFLAGS = -D_CRT_SECURE_NO_DEPRECATE -DCAML_NAME_SPACE
+OCAMLC_CFLAGS = -nologo -O2 -Gy- -MD
+OCAMLC_CPPFLAGS = -D_CRT_SECURE_NO_DEPRECATE
```
It might make sense to put this in the mingw* config files too. Mingw-w64 has a couple references to _CRT_SECURE_NO_DEPRECATE which don't do anything but it might in the future. Since this option is not related to a given compiler but to a platform, it's better to have it in all win32-related places. Moreover there are mix and matches (in particular, some clang setups might MSVC headers).

adrien-n commented on utils/ccomp.ml:
```diff
@@ -82,9 +82,11 @@ let compile_file name =
+                  if !Clflags.native_code
+                  then (Config.ocamlopt_cflags, Config.ocamlopt_cppflags)
+                  else (Config.ocamlc_cflags, Config.ocamlc_cppflags) in
+              (Config.c_compiler ^ " " ^ cflags ^ " " ^ cppflags))
```
I prefer String.concat "" ... here for readability.

adrien-n commented on Makefile:
```diff
@@ -306,8 +306,8 @@ utils/config.ml: utils/config.mlp config/Makefile
 	    -e 's|%%ASM%%|$(ASM)|' \
 	    -e 's|%%ASM_CFI_SUPPORTED%%|$(ASM_CFI_SUPPORTED)|' \
 	    -e 's|%%BYTECCLIBS%%|$(BYTECCLIBS)|' \
-	    -e 's|%%BYTECODE_C_COMPILER%%|$(BYTECODE_C_COMPILER)|' \
```
Do we know if OCaml projects use variables such as this one? If possible, a global grep would be useful. Debian's codesearch gave me no match but I might be using it incorrectly.

adrien-n commented on asmrun/Makefile:
```diff
@@ -37,39 +37,32 @@ LN = ln -s
 ASPPFLAGS = -DSYS_$(SYSTEM)
 ifeq "$(UNIX_OR_WIN32)" "unix"
 ASPPFLAGS += -DMODEL_$(MODEL)
-CFLAGS = $(FLAGS) -g
```
Did we have -g for every build before and now don't have it anymore?

adrien-n reviewed (commented):
As mentioned in another review, I prefer the use of macros rather than variables after which no space must be used because I find them safer to use (it's not possible to put a space where there shouldn't be one).

shindere commented:
adrien-n (2017/03/21 11:08 -0700):
> As mentioned in another review, I prefer the use of macros rather than
> variables after which no space must be used because I find them safer
> to use (it's not possible to put a space where there shouldn't be
> one).

Ah, indeed. Could you show an example of what you had suggested, please?
Can't remember where it was and would take a bit of time to find it
again, sorry.

shindere committed:
configure: export the verbose variable earlier

Before this commit, auxiliary scripts were called before verbose
was exported, so they were not executed in verbose mode.

This commit fixes this and makes sure all the auxiliary scripts take
the verbose environment variable into acocunt.

shindere commented:
adrien-n (2017/03/21 09:43 -0700):
> You can/should write only
> 
>     export cc
> 
> More portable (even though we don't care about systems with shells
> that don't handle this) and consistent.

Very well sppotted, thanks! This is now fixed and another fix has also
been added.

shindere commented:
adrien-n (2017/03/21 10:12 -0700):
> adrien-n commented on this pull request.
> 
> 
> 
> > @@ -951,17 +944,11 @@ case "$native_compiler" in
>        arch=none; model=default; system=unknown; natdynlink=false;;
>  esac
>  
> -if test -z "$ccoption"; then
> 
> To be sure I got it right and for other reviewers, this block is
> removed because it duplicates functionality which is around line 310
> and already sets $cc according to $ccoption. Is tha tright?

Yes. Before this commit there were two C compilers that had to be
defined, namely bytecc and nativecc. This was done that way for
historical reasons but is not necessary any longer.

> > @@ -108,7 +108,7 @@ mylib.cmxa: plugin.cmx plugin2.cmx
>  	@$(OCAMLOPT) $(COMPFLAGS) -a -o mylib.cmxa plugin.cmx plugin2.cmx
>  
>  factorial.$(O): factorial.c
> -	@$(OCAMLOPT) -c -ccopt "$(SHAREDCCCOMPOPTS)" -ccopt -I \
> +	$(OCAMLOPT) -verbose -c -ccopt "$(SHAREDCCCOMPOPTS)" -ccopt -I \
> 
> You've lost the '@' at the beginning of the command-line.

Yeah and the '-verbose' has been added while it should not have been,
thanks! This is now fixed.

shindere commented:
adrien-n (2017/03/21 10:39 -0700):
> It might make sense to put this in the mingw* config files too.
> Mingw-w64 has a couple references to _CRT_SECURE_NO_DEPRECATE which
> don't do anything but it might in the future. Since this option is not
> related to a given compiler but to a platform, it's better to have it
> in all win32-related places. Moreover there are mix and matches (in
> particular, some clang setups might MSVC headers).

I don't know really. See for instance
https://www.quora.com/What-is-the-use-of-define-_CRT_SECURE_NO_DEPRECATE
which says it is Visual studio specific.

So I don't know whether this shouyld be changed or not. And if it
should, I don't know whether it should be in this PR or whether it would
deserve an PR on its own?

shindere commented:
adrien-n (2017/03/21 11:02 -0700):
> adrien-n commented on this pull request.
> 
> 
> 
> > @@ -82,9 +82,11 @@ let compile_file name =
>           (match !Clflags.c_compiler with
>            | Some cc -> cc
>            | None ->
> -              if !Clflags.native_code
> -              then Config.native_c_compiler
> -              else Config.bytecomp_c_compiler)
> +              let (cflags, cppflags) =
> +                  if !Clflags.native_code
> +                  then (Config.ocamlopt_cflags, Config.ocamlopt_cppflags)
> +                  else (Config.ocamlc_cflags, Config.ocamlc_cppflags) in
> +              (Config.c_compiler ^ " " ^ cflags ^ " " ^ cppflags))
> 
> I prefer String.concat "" ... here for readability.

You mean String.concat " ", right?

To be honest, I'm not sure. It may make the code a bit more readable,
indeed, but I am a bit reluctant to build a list just for that. Of
course, I could build a list from the beginning rather than a pair and
then add Config.c_compiler to it but all in all I am not sure I would prefere
the resulting code so at the moment I'd rather leave it as it is. I of
course remain open to comments and may change my mind.

shindere commented:
@adrien-n: as foreseen, I changed my mind and updated utils/ccomp.ml
to take your suggestion into account, after having discussed it with
@gasche. Thanks to both of you.

shindere commented:
adrien-n (2017/03/21 11:06 -0700):
> adrien-n commented on this pull request.
> 
> 
> 
> > @@ -306,8 +306,8 @@ utils/config.ml: utils/config.mlp config/Makefile
>  	    -e 's|%%ASM%%|$(ASM)|' \
>  	    -e 's|%%ASM_CFI_SUPPORTED%%|$(ASM_CFI_SUPPORTED)|' \
>  	    -e 's|%%BYTECCLIBS%%|$(BYTECCLIBS)|' \
> -	    -e 's|%%BYTECODE_C_COMPILER%%|$(BYTECODE_C_COMPILER)|' \
> 
> Do we know if OCaml projects use variables such as this one? If
> possible, a global grep would be useful. Debian's codesearch gave me
> no match but I might be using it incorrectly.

In fact, BYTECODE_C_COMPILER and NATIVE_C_COMPILER have been introduced
very recently (after 4.04) so they do not appear in any release of
OCaml. @gasche suggested to even cherry-pick this PR on 4.05 which may
indeed be a good idea.

shindere commented:
adrien-n (2017/03/21 11:07 -0700):
> adrien-n commented on this pull request.
> 
> 
> 
> >  
>  ASPPFLAGS = -DSYS_$(SYSTEM)
>  ifeq "$(UNIX_OR_WIN32)" "unix"
>  ASPPFLAGS += -DMODEL_$(MODEL)
> -CFLAGS = $(FLAGS) -g
> 
> Did we have -g for every build before and now don't have it anymore?

No. Before this PR, -g was used spuriously in one or two places. I just
systematically got rid of it with the idea that the ongoing changes
will make it even easier to apply it systematically (e.g. with
./configure -ccopt -g).

shindere commented:
@adrien-n: regarding your last remark about OUTPUTOBJ and OUTPUTEXE and the
comment you were referring to
(https://github.com/ocaml/ocaml/pull/981#discussion_r93832174):

The version without space works for both '-o and -Fo (it's what is currently
implemented by this PR). 
However, it is true that the current construct is kind of fragile
and could be broken just by adding a space, which is not the case
with the one you are suggesting.

Still, at the moment I have to say I tend to prefere the current
implementation because, when used, it is slightly shorter.
 
Would it be okay to keep it as it is?

Other opinions are welcome.

shindere committed:
Introduce the LIB_MANEXT build variable

It contains the section for library function manual pages (3).

This will be useful to properly install the manual pages generated
by ocamldoc for the standard library.

shindere commented:
Rebased and added one commit to introduce the LIB_MANEXT build variable.

adrien-n commented:
Sorry for the delay.

Indeed, this is exactly what I had in mind. I leave it to you to decide
what you prefer here as it is only a slight preferene and it's not a
chunk of code that is typically often modified.

adrien-n commented:
I don't find the name very explicit. "SECTION" is also more correct for
manpages.

shindere commented:
adrien-n (2017/03/26 15:04 -0700):
> Sorry for the delay.
> 
> Indeed, this is exactly what I had in mind. I leave it to you to decide
> what you prefer here as it is only a slight preferene and it's not a
> chunk of code that is typically often modified.

Okay, I'll leave it the way it is at the moment, then.

shindere committed:
Build system: rename variables related to manual pages

Rename MANEXT (resp. LIB_MANEXT) to PROGRAMS_MAN_SECTION (resp.
LIBRARIES_MAN_SECTION).

Also get rid of a useless variable assignment in configure.

shindere commented:
adrien-n (2017/03/26 15:05 -0700):
> I don't find the name very explicit. "SECTION" is also more correct for
> manpages.

Yeah, I followed the conservative path.

But, PR rebased on latest trunk and updated to also include a
variables-renaming commit.

damiendoligez commented on configure:
```diff
@@ -354,12 +353,17 @@ esac
 
-bytecc="$cc"
-mkexe="\$(BYTECC)"
+mkexe="\$(CC) \$(CFLAGS) $(CPPFLAGS)"
```
The expansion of `CC` and `CFLAGS` are delayed, but not `CPPFLAGS`. Is this on purpose? If so, you should add a comment.

damiendoligez reviewed (changes requested):
Looks good.

shindere commented:
Damien Doligez (2017/03/27 06:19 -0700):
> > @@ -354,12 +353,17 @@ esac
>  # in the OCaml distribution and third-party C source files compiled
>  # with ocamlc.
>  
> -bytecc="$cc"
> -mkexe="\$(BYTECC)"
> +mkexe="\$(CC) \$(CFLAGS) $(CPPFLAGS)"
> 
> The expansion of `CC` and `CFLAGS` are delayed, but not `CPPFLAGS`. Is
> this on purpose? If so, you should add a comment.

No it was a mistake, thanks for your careful review!

It is now fixed.

shindere commented:
Rebased on latest trunk once more.

Could this be merged now, please?

shindere commented:
Gasp! This PR broke CI on MacOS 10.8 (but works on 10.10).

More specifically, the problem on 10.8 seems to be that the installed
version of ld does not suppport the -ofoo (without space) syntax
for naming executables.

How important is it to fix this?

dra27 commented:
Is it a problem to have the `-o` version have a space after it and the `-Fe` version not?

shindere commented:
David Allsopp (2017/03/29 07:43 -0700):
> Is it a problem to have the `-o` version have a space after it and the
> `-Fe` version not?

Perhaps not, indeed. I'll give it a try and report the result here.

>

dra27 commented:
You know the $(EMPTY) trick for getting a space at the end of the variable in make?

shindere commented:
David Allsopp (2017/03/29 08:16 -0700):
> You know the $(EMPTY) trick for getting a space at the end of the
> variable in make?

Actually no, I don't. I had another idea to add the space but if you
don't mind explaining, I am interested.

dra27 commented:
It's straight from the @damiendoligez make playbook (I think this is correct):

```bash
echo "EMPTY=" >> Makefile
echo "OUTPUTEXE=-o $(EMPTY)" >> Makefile
```

shindere commented:
David Allsopp (2017/03/29 11:43 -0700):
> It's straight from the @damiendoligez make playbook (I think this is correct):

What is this playbook you are mentionning?

> ```bash
> echo "EMPTY=" >> Makefile
> echo "OUTPUTEXE=-o $(EMPTY)" >> Makefile
> ```

Ah as simple as that, I see. OK, thanks!
 
I think I still would like to know whether we really want to keep the
suport for MacOS 10.8, because it is the only machine on which trunk
does not build. And Wikipedia says this version of MacOS has become
unsupported in September 2015.

dra27 commented:
Two thoughts: first, most people when manually invoking gcc will add the space after -o (indeed, I think someone when this was discussed elsewhere was surprised that it even officially worked), so I don't think it's bad to be displaying the space always for *nix platforms (I guess there could be an explicit autoconf test for whether -ofoo works? I probably would, but then that's me...). Second, as the keeper of the Windows museum, I think we should continue to support an old OS version when the changes remain minimal, which I think this is (unless we have evidence of zero users on that platform, of course!).

Is this definitely an OSX 10.8 issue only, or to do with the version of SDKs/packages on that 10.8 installation (what I mean is potentially could someone be using an old version of these tools on a newer and support MacOS installation?).

adrien-n commented:
I think you should keep it because the fix is simple and clean (I didn't
know that trick either btw) and might avoid future bug reports. Not
using a space there also seems to rely on a behaviour that is not
documented, at least according to the manpages on my machine.

dra27 commented:
Just for info, it's a documented feature of GNU getopt (see [here](https://linux.die.net/man/3/getopt)), but of course the gcc driver has not always been getopt-orientated. It may also be worth doing in case there are other obscure cc's out there which can't cope?

xavierleroy commented:
> It may also be worth doing in case there are other obscure cc's out there which can't cope?

As a case in point, CompCert doesn't handle `-o`*filename* without space.  

Note also that the OCaml convention (module Arg, etc) is that options are separated from their arguments by one space.  I think it is a saner convention than getopt's run-together options.  So, please, do not expect the C compiler to implement non-sane conventions.

shindere commented:
Xavier Leroy (2017/03/30 02:08 -0700):
> > It may also be worth doing in case there are other obscure cc's out there which can't cope?
> 
> As a case in point, CompCert doesn't handle `-o`*filename* without space.  
> 
> Note also that the OCaml convention (module Arg, etc) is that options
> are separated from their arguments by one space.  I think it is a
> saner convention than getopt's run-together options.  So, please, do
> not expect the C compiler to implement non-sane conventions.

Fix pushed to trunk in commit dc3038ced643ceafbad046643186aa60ecab8c8c

fdopen commented on utils/config.mlp:
```diff
@@ -135,9 +138,10 @@ let print_config oc =
   p "standard_runtime" standard_runtime;
   p "ccomp_type" ccomp_type;
-  p "bytecomp_c_compiler" bytecomp_c_compiler;
+  p "c_compiler" c_compiler;
```
This breaks all build scripts that parse the output of `ocamlc -config` in order to determine the c compiler - but there is no entry in `Changes` 😒

dra27 replied:
Oops - although how many packages? One could of course argue it reveals how many packages incorrectly parse the output `ocamlc -config` to compile C stubs rather than correctly calling ocamlc/ocamlopt on a C file!

yallop replied:
Yes: for example, the ctypes build is [broken](https://travis-ci.org/ocamllabs/ocaml-ctypes/jobs/279517112#L3259) as a result of this change, since the ctypes makefile [uses the output of `ocamlc -config`](https://github.com/ocamllabs/ocaml-ctypes/blob/8e9fe43ecc4257cfc71a5f744f8708feeeb03b39/Makefile.tests#L7).

gasche replied:
Is it possible to change the relevant build scripts to use `ocamlc` as the C compiler, as @dra27 suggests? If this works, it is both sensible and backward-compatible, so that may be a nice fix. Otherwise, we may want to re-export `{bytecomp,native}_c_compiler` in -config (defined/derived from `c_compiler` and `ocaml{c,opt}_cflags`) for compatibility.

gasche replied:
I created a Mantis issue to discuss this problem. [MPR#7650](https://caml.inria.fr/mantis/view.php?id=7650). Let's move the discussion there instead of lost in a github commit comment.

  https://caml.inria.fr/mantis/view.php?id=7650
